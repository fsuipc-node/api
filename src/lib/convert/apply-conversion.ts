import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
