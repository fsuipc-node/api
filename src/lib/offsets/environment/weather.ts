import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const weather: OffsetList = {
  metarStationAltitude: new Offset({
    value: 0x4B4,
    name: 'metarStationAltitude',
    category: OffsetCategory.WEATHER,
    description: 'METAR station altitude - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  windSurfaceTurbulenceGustSpeed: new Offset({
    value: 0x4BA,
    name: 'windSurfaceTurbulenceGustSpeed',
    category: OffsetCategory.WEATHER,
    description: 'METAR station altitude - kt',
    type: Type.UInt16,
    convert: '+({VAL}).toFixed(2)',
    permission: 'r',
  }),
  metarBarometricDrift: new Offset({
    value: 0x4BC,
    name: 'metarBarometricDrift',
    category: OffsetCategory.WEATHER,
    description: 'METAR barometric drift - difference between aircraft/METAR QNH. Adding drift will give correct value for ATIS report',
    type: Type.UInt16,
    convert: '{VAL} / 16',
    permission: 'r',
  }),
  metarVisibility: new Offset({
    value: 0x4C0,
    name: 'metarVisibility',
    category: OffsetCategory.WEATHER,
    description: 'METAR visibility - sm',
    type: Type.UInt16,
    convert: '{VAL} / 100',
    permission: 'r',
  }),
  metarCloudThunderBase: new Offset({
    value: 0x4C2,
    name: 'metarCloudThunderBase',
    category: OffsetCategory.WEATHER,
    description: 'METAR visibility - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  metarCloudBaseLow: new Offset({
    value: 0x4C4,
    name: 'metarCloudBaseLow',
    category: OffsetCategory.WEATHER,
    description: 'METAR cloud base low - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  metarCloudBaseHigh: new Offset({
    value: 0x4C6,
    name: 'metarCloudBaseHigh',
    category: OffsetCategory.WEATHER,
    description: 'METAR cloud base high - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  dewPoint: new Offset({
    value: 0x4C8,
    name: 'dewPoint',
    category: OffsetCategory.WEATHER,
    description: 'dew point - FS2000+',
    type: Type.Int16,
    convert: '{VAL} / 256',
    permission: 'r',
  }),
  precipitationRate: new Offset({
    value: 0x4C8,
    name: 'precipitationRate',
    category: OffsetCategory.WEATHER,
    description: 'precipitation rate - FS2000+ - 0 to 5',
    type: Type.Byte,
    permission: 'r',
  }),
  precipitationType: new Offset({
    value: 0x4CC,
    name: 'precipitationType',
    category: OffsetCategory.WEATHER,
    description: 'precipitation type - FS2000+ - 0 to 5',
    type: Type.Byte,
    convert: 'precipitationType',
    mapping: true,
    permission: 'r',
  }),
  cloudThunderCoverage: new Offset({
    value: 0x4CD,
    name: 'cloudThunderCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud thunder coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudLowCoverage: new Offset({
    value: 0x4CE,
    name: 'cloudLowCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud low coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudHighCoverage: new Offset({
    value: 0x4CF,
    name: 'cloudHighCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud high coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  precipitationControl: new Offset({
    value: 0x4D2,
    name: 'precipitationControl',
    category: OffsetCategory.WEATHER,
    description: 'hi-byte type 0-2 - low-byte rate 0-5 - writing 0xFFFF release control',
    type: Type.UInt16,
    permission: 'w',
  }),
  dewPointControl: new Offset({
    value: 0x4D4,
    name: 'dewPointControl',
    category: OffsetCategory.WEATHER,
    description: 'degrees / 256 - 0x8000 release control',
    type: Type.UInt16,
    permission: 'w',
  }),
  surfaceWindSpeed: new Offset({
    value: 0x4D8,
    name: 'surfaceWindSpeed',
    category: OffsetCategory.WEATHER,
    description: 'surface wind speed - FS2000+ - kts',
    type: Type.UInt16,
    permission: 'r',
  }),
  surfaceWindDirection: new Offset({
    value: 0x4DA,
    name: 'surfaceWindDirection',
    category: OffsetCategory.WEATHER,
    description: 'surface wind direction - FS2000+ - deg MAG',
    type: Type.UInt16,
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    permission: 'r',
  }),
  unlimitedVisibility: new Offset({
    value: 0xBF8,
    name: 'unlimitedVisibility',
    category: OffsetCategory.WEATHER,
    description: 'unlimited visibility value - sm',
    type: Type.Int32,
    convert: '+({VAL} / 1600).toFixed(2)',
    permission: 'rw',
  }),
  cloudTypeAtAircraftAltitude: new Offset({
    value: 0xE84,
    name: 'cloudTypeAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud type at aircraft atltitude - 0-10',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudCoverageAtAircraftAltitude: new Offset({
    value: 0xE85,
    name: 'cloudCoverageAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud coverage at aircraft atltitude - okta',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudIcingAtAircraftAltitude: new Offset({
    value: 0xE86,
    name: 'cloudIcingAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud icing at aircraft atltitude - 0-4',
    type: Type.UInt16,
    permission: 'r',
  }),
  cloudTurbulenceAtAircraftAltitude: new Offset({
    value: 0xE88,
    name: 'cloudTurbulenceAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud turbulence at aircraft atltitude - 0-255',
    type: Type.UInt16,
    permission: 'r',
  }),
  visibility: new Offset({
    value: 0xE8A,
    name: 'visibility',
    category: OffsetCategory.WEATHER,
    description: 'current visibility - sm',
    convert: '+({VAL} / 100).toFixed(2)',
    type: Type.Int16,
    permission: 'r',
  }),
  OAT: new Offset({
    value: 0xE8C,
    name: 'OAT',
    category: OffsetCategory.WEATHER,
    description: 'current outside air temperature (OAT) - celsius',
    convert: 'Math.round({VAL} / 256)',
    type: Type.Int16,
    permission: 'r',
  }),
  interpolatedDewPoint: new Offset({
    value: 0xE8E,
    name: 'interpolatedDewPoint',
    category: OffsetCategory.WEATHER,
    description: 'dew point interpolated for aircraft altitude - celsius',
    convert: 'Math.round({VAL} / 256)',
    type: Type.Int16,
    permission: 'r',
  }),
  windSpeed: new Offset({
    value: 0xE90,
    name: 'windSpeed',
    category: OffsetCategory.WEATHER,
    description: 'wind speed at aircraft - kt',
    type: Type.UInt16,
    permission: 'r',
  }),
  windDirection: new Offset({
    value: 0xE92,
    name: 'windDirection',
    category: OffsetCategory.WEATHER,
    description: 'wind direction at aircraft - TRUE degrees',
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    type: Type.UInt16,
    permission: 'r',
  }),
};
