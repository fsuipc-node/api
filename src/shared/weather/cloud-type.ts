export enum CloudType {
  USER = 'user',
  CIRRUS = 'cirrus',
  STRATUS = 'stratus',
  CUMULUS = 'cumulus',
}
