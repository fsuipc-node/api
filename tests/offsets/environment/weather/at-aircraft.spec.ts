import { weatherAtAircraft as offsets } from '@offsets/environment/weather/at-aircraft';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - environment/weather/at-aircraft', () => {
  const offsetsTestCases = [
    { name: 'upperCloudCeiling', value: 914, expectedResult: 2998.69 },
    { name: 'upperCloudBase', value: 914, expectedResult: 2998.69 },
    { name: 'upperCloudCoverage', value: 32768 , expectedResult: 4 },
    { name: 'upperCloudVariation', value: 914, expectedResult: 2998.69 },
    { name: 'lowerCloudCeiling', value: 914, expectedResult: 2998.69 },
    { name: 'lowerCloudBase', value: 914, expectedResult: 2998.69 },
    { name: 'lowerCloudCoverage', value: 32768 , expectedResult: 4 },
    { name: 'lowerCloudVariation', value: 914, expectedResult: 2998.69 },
    { name: 'stormCloudCeiling', value: 914, expectedResult: 2998.69 },
    { name: 'stormCloudBase', value: 914, expectedResult: 2998.69 },
    { name: 'stormCloudCoverage', value: 32768 , expectedResult: 4 },
    { name: 'stormCloudVariation', value: 914, expectedResult: 2998.69 },
    { name: 'upperTemperatureLevel', value: 914, expectedResult: 2998.69 },
    { name: 'upperTemperature', value: 512, expectedResult: 2 },
    { name: 'middleTemperatureLevel', value: 914, expectedResult: 2998.69 },
    { name: 'middleTemperature', value: 512, expectedResult: 2 },
    { name: 'lowerTemperatureLevel', value: 914, expectedResult: 2998.69 },
    { name: 'lowerTemperature', value: 512, expectedResult: 2 },
    { name: 'surfaceTemperatureLevel', value: 914, expectedResult: 2998.69 },
    { name: 'surfaceTemperature', value: 512, expectedResult: 2 },
    { name: 'temperatureDrift', value: 512, expectedResult: 2 },
    { name: 'temperatureDayNightVariation', value: 512, expectedResult: 2 },
    { name: 'pressureQNH', value: 32, expectedResult: 2 },
    { name: 'pressureDrift', value: 32, expectedResult: 2 },
    { name: 'upperWindCeiling', value: 914, expectedResult: 2998.69 },
    { name: 'upperWindBase', value: 914, expectedResult: 2998.69 },
    { name: 'upperWindDirection', value: 32768, expectedResult: 180 },
    { name: 'upperWindGustEnabled', value: 1, expectedResult: true },
    { name: 'upperWindGustEnabled', value: 0, expectedResult: false },
    { name: 'middleWindCeiling', value: 914, expectedResult: 2998.69 },
    { name: 'middleWindBase', value: 914, expectedResult: 2998.69 },
    { name: 'middleWindDirection', value: 32768, expectedResult: 180 },
    { name: 'middleWindGustEnabled', value: 1, expectedResult: true },
    { name: 'middleWindGustEnabled', value: 0, expectedResult: false },
    { name: 'lowerWindCeiling', value: 914, expectedResult: 2998.69 },
    { name: 'lowerWindBase', value: 914, expectedResult: 2998.69 },
    { name: 'lowerWindDirection', value: 32768, expectedResult: 180 },
    { name: 'lowerWindGustEnabled', value: 1, expectedResult: true },
    { name: 'lowerWindGustEnabled', value: 0, expectedResult: false },
    { name: 'surfaceWindCeiling', value: 914, expectedResult: 2998.69 },
    { name: 'surfaceWindGustEnabled', value: 1, expectedResult: true },
    { name: 'surfaceWindGustEnabled', value: 0, expectedResult: false },
    { name: 'upperCloudIcing', value: 1, expectedResult: true },
    { name: 'upperCloudIcing', value: 0, expectedResult: false },
    { name: 'lowerCloudIcing', value: 1, expectedResult: true },
    { name: 'lowerCloudIcing', value: 0, expectedResult: false },
    { name: 'stormCloudIcing', value: 1, expectedResult: true },
    { name: 'stormCloudIcing', value: 0, expectedResult: false },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
