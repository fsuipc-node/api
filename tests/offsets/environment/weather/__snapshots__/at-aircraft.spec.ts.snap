// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - environment/weather/at-aircraft offsets list should have required properties 1`] = `
Object {
  "ambientAirDensity": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "ambient air density - slug / sq ft",
    "length": undefined,
    "mapping": undefined,
    "name": "ambientAirDensity",
    "permission": "r",
    "type": 8,
    "value": 10432,
  },
  "ambientAirPressure": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "ambient air pressure - lbs / sq ft",
    "length": undefined,
    "mapping": undefined,
    "name": "ambientAirPressure",
    "permission": "r",
    "type": 8,
    "value": 10440,
  },
  "lowerCloudBase": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower cloud layer base - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudBase",
    "permission": "r",
    "type": 5,
    "value": 3748,
  },
  "lowerCloudCeiling": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower cloud layer ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudCeiling",
    "permission": "r",
    "type": 5,
    "value": 3746,
  },
  "lowerCloudCoverage": Offset {
    "category": "weather",
    "convert": "Math.round(({VAL} + 1) / 8192)",
    "description": "lower cloud layer coverage - oktas",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudCoverage",
    "permission": "r",
    "type": 5,
    "value": 3750,
  },
  "lowerCloudIcing": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "lower cloud icing",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudIcing",
    "permission": "r",
    "type": 5,
    "value": 3840,
  },
  "lowerCloudTurbulence": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "lower cloud turbulence - 0-255",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudTurbulence",
    "permission": "r",
    "type": 5,
    "value": 3842,
  },
  "lowerCloudType": Offset {
    "category": "weather",
    "convert": "cloudType",
    "description": "lower cloud type",
    "length": undefined,
    "mapping": true,
    "name": "lowerCloudType",
    "permission": "r",
    "type": 5,
    "value": 3838,
  },
  "lowerCloudVariation": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower cloud layer variation - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudVariation",
    "permission": "r",
    "type": 5,
    "value": 3752,
  },
  "lowerTemperature": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "lower temperature - degrees C",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerTemperature",
    "permission": "r",
    "type": 2,
    "value": 3772,
  },
  "lowerTemperatureLevel": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower temperature level - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerTemperatureLevel",
    "permission": "r",
    "type": 5,
    "value": 3770,
  },
  "lowerWindBase": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower wind ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindBase",
    "permission": "r",
    "type": 5,
    "value": 3812,
  },
  "lowerWindCeiling": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower wind ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindCeiling",
    "permission": "r",
    "type": 5,
    "value": 3810,
  },
  "lowerWindDirection": Offset {
    "category": "weather",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "lower wind ceiling - TRUE degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindDirection",
    "permission": "r",
    "type": 5,
    "value": 3816,
  },
  "lowerWindGustEnabled": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "lower wind gust enabled",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindGustEnabled",
    "permission": "r",
    "type": 2,
    "value": 3820,
  },
  "lowerWindSpeed": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "lower wind ceiling - kt",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindSpeed",
    "permission": "r",
    "type": 5,
    "value": 3814,
  },
  "lowerWindTurbulence": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "lower wind turbulence - 0-255",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindTurbulence",
    "permission": "r",
    "type": 5,
    "value": 3818,
  },
  "middleTemperature": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "middle temperature - degrees C",
    "length": undefined,
    "mapping": undefined,
    "name": "middleTemperature",
    "permission": "r",
    "type": 2,
    "value": 3768,
  },
  "middleTemperatureLevel": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "middle temperature level - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "middleTemperatureLevel",
    "permission": "r",
    "type": 5,
    "value": 3766,
  },
  "middleWindBase": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "middle wind ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindBase",
    "permission": "r",
    "type": 5,
    "value": 3800,
  },
  "middleWindCeiling": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "middle wind ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindCeiling",
    "permission": "r",
    "type": 5,
    "value": 3798,
  },
  "middleWindDirection": Offset {
    "category": "weather",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "middle wind ceiling - TRUE degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindDirection",
    "permission": "r",
    "type": 5,
    "value": 3804,
  },
  "middleWindGustEnabled": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "middle wind gust enabled",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindGustEnabled",
    "permission": "r",
    "type": 5,
    "value": 3808,
  },
  "middleWindSpeed": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "middle wind ceiling - kt",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindSpeed",
    "permission": "r",
    "type": 5,
    "value": 3802,
  },
  "middleWindTurbulence": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "middle wind turbulence - 0-255",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindTurbulence",
    "permission": "r",
    "type": 5,
    "value": 3806,
  },
  "pressureDrift": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 16)",
    "description": "pressure drift - mb",
    "length": undefined,
    "mapping": undefined,
    "name": "pressureDrift",
    "permission": "r",
    "type": 2,
    "value": 3784,
  },
  "pressureQNH": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 16)",
    "description": "pressure QNH - mb",
    "length": undefined,
    "mapping": undefined,
    "name": "pressureQNH",
    "permission": "r",
    "type": 5,
    "value": 3782,
  },
  "standardDensityRatio": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "sigma - standard density ratio",
    "length": undefined,
    "mapping": undefined,
    "name": "standardDensityRatio",
    "permission": "r",
    "type": 8,
    "value": 10480,
  },
  "standardPressureRatio": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "delta - standard pressure ratio",
    "length": undefined,
    "mapping": undefined,
    "name": "standardPressureRatio",
    "permission": "r",
    "type": 8,
    "value": 10472,
  },
  "standardTemperatureRatio": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "theta - standard temperature ratio",
    "length": undefined,
    "mapping": undefined,
    "name": "standardTemperatureRatio",
    "permission": "r",
    "type": 8,
    "value": 10464,
  },
  "staticAirTemperatureF": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "static air temperature - F",
    "length": undefined,
    "mapping": undefined,
    "name": "staticAirTemperatureF",
    "permission": "r",
    "type": 8,
    "value": 10448,
  },
  "staticAirTemperatureR": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "static air temperature - Rankine degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "staticAirTemperatureR",
    "permission": "r",
    "type": 8,
    "value": 10456,
  },
  "stormCloudBase": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "storm cloud layer base - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudBase",
    "permission": "r",
    "type": 5,
    "value": 3756,
  },
  "stormCloudCeiling": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "storm cloud layer ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudCeiling",
    "permission": "r",
    "type": 5,
    "value": 3754,
  },
  "stormCloudCoverage": Offset {
    "category": "weather",
    "convert": "Math.round(({VAL} + 1) / 8192)",
    "description": "storm cloud layer coverage - oktas",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudCoverage",
    "permission": "r",
    "type": 5,
    "value": 3758,
  },
  "stormCloudIcing": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "storm cloud icing",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudIcing",
    "permission": "r",
    "type": 5,
    "value": 3846,
  },
  "stormCloudTurbulence": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "storm cloud turbulence - 0-255",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudTurbulence",
    "permission": "r",
    "type": 5,
    "value": 3848,
  },
  "stormCloudType": Offset {
    "category": "weather",
    "convert": "cloudType",
    "description": "storm cloud type",
    "length": undefined,
    "mapping": true,
    "name": "stormCloudType",
    "permission": "r",
    "type": 5,
    "value": 3844,
  },
  "stormCloudVariation": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "storm cloud layer variation - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudVariation",
    "permission": "r",
    "type": 5,
    "value": 3760,
  },
  "surfaceTemperature": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "surface temperature - degrees C",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceTemperature",
    "permission": "r",
    "type": 2,
    "value": 3776,
  },
  "surfaceTemperatureLevel": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "surface temperature level - ground elevation - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceTemperatureLevel",
    "permission": "r",
    "type": 5,
    "value": 3774,
  },
  "surfaceWindCeiling": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "surface wind ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceWindCeiling",
    "permission": "r",
    "type": 5,
    "value": 3822,
  },
  "surfaceWindGustEnabled": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "surface wind gust enabled",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceWindGustEnabled",
    "permission": "r",
    "type": 5,
    "value": 3830,
  },
  "surfaceWindTurbulence": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "surface wind turbulence - 0-255",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceWindTurbulence",
    "permission": "r",
    "type": 5,
    "value": 3828,
  },
  "temperatureDayNightVariation": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "temperature day night variation - degrees C",
    "length": undefined,
    "mapping": undefined,
    "name": "temperatureDayNightVariation",
    "permission": "r",
    "type": 2,
    "value": 3780,
  },
  "temperatureDrift": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "temperature drift - degrees C",
    "length": undefined,
    "mapping": undefined,
    "name": "temperatureDrift",
    "permission": "r",
    "type": 2,
    "value": 3778,
  },
  "upperCloudBase": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper cloud layer base - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudBase",
    "permission": "r",
    "type": 5,
    "value": 3740,
  },
  "upperCloudCeiling": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper cloud layer ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudCeiling",
    "permission": "r",
    "type": 5,
    "value": 3738,
  },
  "upperCloudCoverage": Offset {
    "category": "weather",
    "convert": "Math.round(({VAL} + 1) / 8192)",
    "description": "upper cloud layer coverage - oktas",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudCoverage",
    "permission": "r",
    "type": 5,
    "value": 3742,
  },
  "upperCloudIcing": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "upper cloud icing",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudIcing",
    "permission": "r",
    "type": 5,
    "value": 3834,
  },
  "upperCloudTurbulence": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "upper cloud turbulence - 0-255",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudTurbulence",
    "permission": "r",
    "type": 5,
    "value": 3836,
  },
  "upperCloudType": Offset {
    "category": "weather",
    "convert": "cloudType",
    "description": "upper cloud type",
    "length": undefined,
    "mapping": true,
    "name": "upperCloudType",
    "permission": "r",
    "type": 5,
    "value": 3832,
  },
  "upperCloudVariation": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper cloud layer variation - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudVariation",
    "permission": "r",
    "type": 5,
    "value": 3744,
  },
  "upperTemperature": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "upper temperature - degrees C",
    "length": undefined,
    "mapping": undefined,
    "name": "upperTemperature",
    "permission": "r",
    "type": 2,
    "value": 3764,
  },
  "upperTemperatureLevel": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper temperature level - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "upperTemperatureLevel",
    "permission": "r",
    "type": 5,
    "value": 3762,
  },
  "upperWindBase": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper wind ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindBase",
    "permission": "r",
    "type": 5,
    "value": 3788,
  },
  "upperWindCeiling": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper wind ceiling - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindCeiling",
    "permission": "r",
    "type": 5,
    "value": 3786,
  },
  "upperWindDirection": Offset {
    "category": "weather",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "upper wind ceiling - TRUE degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindDirection",
    "permission": "r",
    "type": 5,
    "value": 3792,
  },
  "upperWindGustEnabled": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "upper wind gust enabled",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindGustEnabled",
    "permission": "r",
    "type": 5,
    "value": 3796,
  },
  "upperWindSpeed": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "upper wind ceiling - kt",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindSpeed",
    "permission": "r",
    "type": 5,
    "value": 3790,
  },
  "upperWindTurbulence": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "upper wind turbulence - 0-255",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindTurbulence",
    "permission": "r",
    "type": 5,
    "value": 3794,
  },
}
`;
